### DISCUSSION TITLE
Media Controls for Video Player

### DISCUSSION THREAD

**Discussion by esegebart** [react-native-tvos/react-native-tvos#340] [Q&A] [answered]
I am using React Native Video as the video object in my project. I am wondering what kind of media controls/video controls work best with React Native tvOS? I have the Android native controls included with it and I am not able to bring them back up with my remote or what is the best way to go about adding controls? Play, pause, forward, rewind at minimum.

**Comment by douglowder** [comment]
I have a video demo but it is a bit old -- I'll see if I can modernize that and post it here

**Comment by douglowder** [comment]
Here's the demo component that shows most of the controls that worked as of `react-native-video` 5.2.0. : https://github.com/douglowder/RNAppleTVTalk/blob/master/js/VideoDemo.js

Basically all the Video props map to state variables, and the controls are all implemented as RN controls, not native controls.

I need to extract that and redo it as a modern functional component, but it shows you what is possible.

**Reply by esegebart** [reply]
Thank you so much! So just so I am following - I would have all of my play, pause, rewind, as state variables and set them that way and I would basically have custom controls correct? I am looking at renderNativeSkin and renderCustomSkin, and I am trying to understand what the difference is between the two. Is one for phone and one for the TV? CustomSkin is the TV? I've been doing React Native for about a year and still learning a lot.  I think I kind of see how I can do this. :) Thanks!

**Reply by douglowder** [reply]
That is definitely one way to do it, with state variables. I'll have a look at this example -- I'm not sure what the difference in skins is, and it seems to me the example could be simplified a LOT :)

**Reply by esegebart** [reply]
Yes, I am still trying to figure out the controls myself. I was a little confused by the tvEventHandler and what it means - was trying to look it up and get a clearer picture. I have only been working with React Native for about a year so a TV app is a pretty large endeavor for me. LOL. When you say the controls are all implemented at RN controls - what do you mean by that?

I see a lot of examples of custom controls for React Native Video - but then I worry it will not work for tvOS/firestick/androidTV.  I thought about making an overlay and make my own buttons and then just use the props/functions with RN Video, but then I don't know how the overlay will be triggered since it is triggered with touch.  

Like on netflix or hbomax - it seems like they have some sort of overlay component when the tv remote hits a button - that's what I am trying to figure out.

**Comment by freeboub** [comment]
Hello, I think this ticket is more related to react-native-video :) 
Anyway it seems that following project is the most used with react-native-video:
https://github.com/itsnubix/react-native-video-controls
But i don't know if focus works on TV

The issue you describe (not focusable controls on TV) is an issue in react-native-video! But as there is no full time maintainer you will need to fix it ...

The issue with built in controls is that they are not unified across platforms, I also think a highly customizable component/layer is useful.

**Reply by douglowder** [reply]
Thanks @freeboub I'll have a look at that and see what the level of effort is to make those controls work on TV

**Reply by esegebart** [reply]
@freeboub I guess I should clarify - I am not asking what controls will work with React Native Video. I am asking what is the best route to make controls that this fork for tvOS can use and be used with the remote? Or do I have to go completely custom?

**Comment by douglowder** [comment, marked as answer]
@esegebart here is an updated video demo showing some basics of how to control playback, see progress, etc. It's a lot more modern in style than the older one above.

https://github.com/douglowder/TVInputDemo/blob/master/src/screens/VideoExample.tsx

**Reply by esegebart** [reply]
Amazing! Thank you so much Mr. Lowder!!! Can't wait to get it working in my project tomorrow!

**Comment by LunatiqueCoder** [comment]
Hello everybody!

Thanks to @esegebart I added minimal tvOS support to [react-native-media-console](https://github.com/criszz77/react-native-media-console) which is basically rewritten completely from [react-native-video-controls](https://github.com/itsnubix/react-native-video-controls) with TypeScript, I solved all the issues I could find and now we're testing the tvOS support on `react-native-media-console@next`. Should be merged by tomorrow if everything goes well. :)

**Reply by LunatiqueCoder** [reply]
@esegebart It should work both with tvOS and without. 

If you find any issues, please document them on my repo 🙏

**Reply by esegebart** [reply]
Yes, thanks to @criszz77 we have an amazing video and controls component that will work with the remote so if you are looking for a highly customizable media controls - this is the total package!!!

**Comment by freeboub** [comment]
Subtitles and audio tracks support seems missing ;) anyway the UI looks good !

**Reply by LunatiqueCoder** [reply]
@freeboub Thank you! I haven't worked with tvOS at all, but I'll create an issue with this and have a look into it. Feel free to also open issues if you find other problems or if you have suggestions.
